server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1131,72 +1430,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
+                        // Failure due to not being master any more so we don't try to do run more cluster state updates. The next master
+                        // will try handling the missing operations. All we can do is fail all the listeners on this master node so that
+                        // transport requests return and we don't leak listeners
+                        final Exception wrapped =
+                                new RepositoryException(repoName, "Failed to update cluster state during snapshot delete", e);
+                        final Deque<SnapshotFinalization> outstandingSnapshotsForRepo = snapshotsToFinalize.remove(repoName);
+                        if (outstandingSnapshotsForRepo != null) {
+                            SnapshotFinalization finalization;
+                            while ((finalization = outstandingSnapshotsForRepo.poll()) != null) {
+                                failSnapshotCompletionListeners(finalization.entry.snapshot(), wrapped);
+                            }
+                        }
+                        for (Iterator<List<ActionListener<RepositoryData>>> iterator = snapshotDeletionListeners.values().iterator();
+                             iterator.hasNext(); ) {
+                            List<ActionListener<RepositoryData>> listeners = iterator.next();
+                            iterator.remove();
+                            failListenersIgnoringException(listeners, wrapped);
+                        }
+                        assert snapshotDeletionListeners.isEmpty() :
+                                "No new listeners should have been added but saw " + snapshotDeletionListeners;
+                    } else {
+                        assert false : "Removing snapshot entry should only ever fail because we failed to publish new state";
+                    }
+                    currentlyFinalizing.remove(repoName);
                 }
             }
 
             @Override
             public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
-                if (listener != null) {
-                    if (failure != null) {
-                        listener.onFailure(failure);
-                    } else {
-                        logger.info("Successfully deleted snapshots {}", snapshotIds);
-                        listener.onResponse(null);
+                final List<ActionListener<RepositoryData>> deleteListeners = snapshotDeletionListeners.remove(deleteEntry.uuid());
+                if (failure == null) {
+                    completeListenersIgnoringException(deleteListeners, repositoryData);
+                } else {
+                    failListenersIgnoringException(deleteListeners, failure);
+                }
+                runningDeletions.remove(deleteEntry.uuid());
+                if (repositoryData == null) {
```

Comment:
how is this tested?

Reply:
`org.elasticsearch.snapshots.ConcurrentSnapshotsIT#testQueuedOperationsAndBrokenRepoOnMasterFailOver` reproduces this scenario by corrupting the index-N blob before master fail-over.